File: src/main/java/ch/itds/taglib/asset/pipeline/thymeleaf/AssetDialect.java
package ch.itds.taglib.asset.pipeline.thymeleaf;

import org.thymeleaf.dialect.AbstractProcessorDialect;
import org.thymeleaf.dialect.IExpressionObjectDialect;
import org.thymeleaf.expression.IExpressionObjectFactory;
import org.thymeleaf.processor.IProcessor;
import org.thymeleaf.standard.StandardDialect;
import org.thymeleaf.standard.processor.StandardXmlNsTagProcessor;
import org.thymeleaf.templatemode.TemplateMode;

import java.util.HashSet;
import java.util.Set;

public class AssetDialect extends AbstractProcessorDialect implements IExpressionObjectDialect {

    private static final String DIALECT_NAME_SPACE = "http://www.itds.ch/taglib/assets";

    private final transient IExpressionObjectFactory EXPRESSION_OBJECTS_FACTORY;

    public AssetDialect() {
        super(DIALECT_NAME_SPACE, "asset", StandardDialect.PROCESSOR_PRECEDENCE);
        EXPRESSION_OBJECTS_FACTORY = new AssetExpressionFactory();
    }

    public Set<IProcessor> getProcessors(final String dialectPrefix) {
        final Set<IProcessor> processors = new HashSet<>();
        processors.add(new AssetHrefAttributeTagProcessor(dialectPrefix));
        processors.add(new AssetSrcAttributeTagProcessor(dialectPrefix));
        processors.add(new StandardXmlNsTagProcessor(TemplateMode.HTML, dialectPrefix));
        return processors;
    }

    @Override
    public IExpressionObjectFactory getExpressionObjectFactory() {
        return EXPRESSION_OBJECTS_FACTORY;
    }
}

File: src/main/java/com/dra11y/assetThymeleafTaglib/AssetDialect.kt
package com.dra11y.assetThymeleafTaglib

import com.dra11y.assetThymeleafTaglib.asset.AssetResolver
import com.dra11y.assetThymeleafTaglib.expression.AssetExpressionFactory
import com.dra11y.assetThymeleafTaglib.processor.AssetHrefAttributeTagProcessor
import com.dra11y.assetThymeleafTaglib.processor.AssetSrcAttributeTagProcessor
import org.springframework.stereotype.Component
import org.thymeleaf.dialect.AbstractProcessorDialect
import org.thymeleaf.dialect.IExpressionObjectDialect
import org.thymeleaf.processor.IProcessor
import org.thymeleaf.standard.StandardDialect
import org.thymeleaf.standard.processor.StandardXmlNsTagProcessor
import org.thymeleaf.templatemode.TemplateMode

@Component
open class AssetDialect(private val assetExpressionFactory: AssetExpressionFactory, private val assetResolver: AssetResolver) :
    AbstractProcessorDialect(DIALECT_NAMESPACE, "asset", StandardDialect.PROCESSOR_PRECEDENCE),
    IExpressionObjectDialect {

    override fun getProcessors(dialectPrefix: String): Set<IProcessor> {
        val processors: MutableSet<IProcessor> = HashSet()
        processors.add(AssetHrefAttributeTagProcessor(dialectPrefix, assetResolver))
        processors.add(AssetSrcAttributeTagProcessor(dialectPrefix, assetResolver))
        processors.add(StandardXmlNsTagProcessor(TemplateMode.HTML, dialectPrefix))
        return processors
    }

    override fun getExpressionObjectFactory() = assetExpressionFactory

    companion object {
        private const val DIALECT_NAMESPACE = "https://dra11y.github.io/asset-thymeleaf-taglib-kotlin"
    }
}
